pkg/internal/apis/certmanager/validation/certificaterequest_test.go

```diff
@@ -100,136 +105,651 @@ func TestValidateCertificateRequestUpdate(t *testing.T) {
+					Conditions: []cminternal.CertificateRequestCondition{
+						{
+							Type:   cminternal.CertificateRequestConditionApproved,
+							Status: cminternalmeta.ConditionTrue,
+							Reason: "cert-manager.io",
+						},
+					},
+				},
+			},
+			want: nil,
+		},
+		"CertificateRequest with no condition that changes to Denied=true, shouldn't error": {
+			oldCR: &cminternal.CertificateRequest{
+				Spec: cminternal.CertificateRequestSpec{
+					Request:   baseRequest,
+					IssuerRef: validIssuerRef,
+				},
+				Status: cminternal.CertificateRequestStatus{
+					Conditions: []cminternal.CertificateRequestCondition{},
+				},
+			},
+			newCR: &cminternal.CertificateRequest{
+				Spec: cminternal.CertificateRequestSpec{
+					Request:   baseRequest,
+					IssuerRef: validIssuerRef,
+				},
+				Status: cminternal.CertificateRequestStatus{
+					Conditions: []cminternal.CertificateRequestCondition{
+						{
+							Type:   cminternal.CertificateRequestConditionDenied,
+							Status: cminternalmeta.ConditionTrue,
+							Reason: "Foo",
+						},
+					},
+				},
+			},
+			want: nil,
+		},
+		"CertificateRequest with single Approved=true condition that is removed, should error": {
+			oldCR: &cminternal.CertificateRequest{
+				Spec: cminternal.CertificateRequestSpec{
+					Request:   baseRequest,
+					IssuerRef: validIssuerRef,
+				},
+				Status: cminternal.CertificateRequestStatus{
+					Conditions: []cminternal.CertificateRequestCondition{
+						{
+							Type:   cminternal.CertificateRequestConditionApproved,
+							Status: cminternalmeta.ConditionTrue,
+						},
+					},
+				},
+			},
+			newCR: &cminternal.CertificateRequest{
+				Spec: cminternal.CertificateRequestSpec{
+					Request:   baseRequest,
+					IssuerRef: validIssuerRef,
+				},
+				Status: cminternal.CertificateRequestStatus{
+					Conditions: []cminternal.CertificateRequestCondition{},
+				},
+			},
+			want: []*field.Error{
+				field.Forbidden(fldPathConditions, "'Approved' condition may not be modified once set"),
+			},
+		},
+		"CertificateRequest with single Denied=true condition that is removed, should error": {
+			oldCR: &cminternal.CertificateRequest{
+				Spec: cminternal.CertificateRequestSpec{
+					Request:   baseRequest,
+					IssuerRef: validIssuerRef,
+				},
+				Status: cminternal.CertificateRequestStatus{
+					Conditions: []cminternal.CertificateRequestCondition{
+						{
+							Type:   cminternal.CertificateRequestConditionDenied,
+							Status: cminternalmeta.ConditionTrue,
+						},
+					},
+				},
+			},
+			newCR: &cminternal.CertificateRequest{
+				Spec: cminternal.CertificateRequestSpec{
+					Request:   baseRequest,
+					IssuerRef: validIssuerRef,
+				},
+				Status: cminternal.CertificateRequestStatus{
+					Conditions: []cminternal.CertificateRequestCondition{},
+				},
+			},
+			want: []*field.Error{
+				field.Forbidden(fldPathConditions, "'Denied' condition may not be modified once set"),
+			},
+		},
```

Comment:
I wondered whether it'd be worth adding tests for updating the last modified time, or the observed generation fields of an existing condition.

Reply:
We don't have observed generation for CRs since the spec is immutable so the number will always be 1. Asserting the modified time _could_ be something we check in another PR, though we should introduce that check everywhere.